solara/components/input.py

```diff
@@ -276,113 +288,70 @@ def Page():
+    # Workaround to hook() getting a stale string_value
+    string_value_ref = solara.use_ref(string_value)
+    string_value_ref.current = string_value
 
-    error, set_error = solara.use_state(False)
-    internal_value, set_internal_value = solara.use_state(cast(Union[str, T, None], str(reactive_value.value) if reactive_value.value is not None else None))
+    error_message = cast(Union[str, None], None)
 
-    def parse(value):
-        return ast.literal_eval(value.replace(",", "."))
+    try:
+        reactive_value.set(parse(string_value))
+    except ValueError as e:
+        error_message = str(e.args[0])
 
-    def on_external_value_change():
-        nonlocal internal_value
-        value = reactive_value.value
-        # only gets called when initial_or_updated changes
-        if isinstance(internal_value, str):
-            try:
-                numerical_value = ast.literal_eval(internal_value.replace(",", "."))
-            except Exception:
-                set_internal_value(value)
-                # this make sure the current value gets updated directly
-                internal_value = str(value)
-            else:
-                if numerical_value != value:
-                    set_internal_value(str(value))
-                    # this make sure the current value gets updated directly
-                    internal_value = str(value)
-        elif internal_value is not None:
-            if internal_value != value:
-                set_internal_value(value)
-                # this make sure the current value gets updated directly
-                internal_value = value
-        elif internal_value is None:
-            set_internal_value(value)
-            # this make sure the current value gets updated directly
-            internal_value = value
-
-    # make sure that if the external value changes, our internal model gets updated
-    # but if the internal model is in a different representation
-    # (e.g. internal_value='1e2' and internal_value=100) we don't want to change
-    # our internal model
-    solara.use_memo(on_external_value_change, [reactive_value.value])
-
-    def internal_value_check_type():
-        if isinstance(internal_value, str):
-            try:
-                numerical_proper_type = str_to_numeric(parse(internal_value))
-                numerical = parse(internal_value)
-            except Exception:
-                return internal_value
-            else:
-                if numerical_proper_type != numerical:
-                    return str(numerical_proper_type)
-                else:
-                    return internal_value
-        else:
-            return internal_value
+    def hook():
+        def on_external_value_change(new_value: Optional[T]):
+            new_string_value = stringify(new_value)
+            if new_value != parse(string_value_ref.current):
```

Comment:
Lets add a test to cover a possible problem here:
Enter invalid input, e.g. "1a", not change the reactive variable we passed (e.g. we choose to modify a slider that is linked to the same reactive var). Now this triggers an exception.

Reply:
Fixed and tested for by new commit.